File: src/1024.视频拼接.ts
/*
 * @lc app=leetcode.cn id=1024 lang=typescript
 *
 * [1024] 视频拼接
 */

// @lc code=start
function videoStitching(clips: number[][], T: number): number {
    const maxn = new Array<number>(T).fill(0);
    let last = 0, ret = 0, pre = 0;
    for (const clip of clips) {
        if (clip[0] < T) {
            maxn[clip[0]] = Math.max(maxn[clip[0]], clip[1]);
        }
    }
    for (let i = 0; i < T; i++) {
        last = Math.max(last, maxn[i]);
        if (i == last) {
            return -1;
        }
        if (i == pre) {
            ret++;
            pre = last;
        }
    }
    return ret;
};
// @lc code=end

File: src/1024.视频拼接.go
/*
 * @lc app=leetcode.cn id=1024 lang=golang
 *
 * [1024] 视频拼接
 */
package leetcode

import "math"

// @lc code=start
func videoStitching(clips [][]int, T int) int {
	dp := make([]int, T+1)

	const inf = math.MaxInt64 - 1

	for i := range dp {
		dp[i] = inf
	}

	// base case
	dp[0] = 0
	for i := 1; i < T+1; i++ {
		for _, c := range clips {
			l, r := c[0], c[1]
			if l < i && i <= r && dp[l]+1 < dp[i] {
				dp[i] = dp[l] + 1
			}
		}
	}

	if dp[T] == inf {
		return -1
	}

	return dp[T]
}

// @lc code=end
